### ISSUE TITLE
New settings `clear` and `default` for controlling inheritance of cumulative configs

### ISSUE THREAD

**Issue by brillout** [vikejs/vike#1692]
Add two new settings `default: boolean` and `clear: boolean` for controlling inheritance of cumulative values:

```js
// pages/+config.js

import { Layout } from './Layout'

export default {
  Layout: {
    value: Layout,
    // If a child defines a layout it will replace this layout
    default: true
  }
}
```

```js
// pages/admin-panel/+config.js

import { Layout } from './Layout'

export default {
  // Overrides all previously defined layouts (i.e. all layouts defined by ancestors)
  Layout: {
    value: Layout,
    clear: true
  }
}
```

### Groups

```js
// pages/+config.js

import HeadDefaultOpenGraph from './HeadDefaultOpenGraph'
import HeadGlobal from './HeadGlobal'

export default {
  Head: [
    {
      value: HeadGlobal
    },
    {
      value: HeadDefaultOpenGraph,
      group: 'open-graph' // Can be any string (it's used as a unique identifying key)
    }
  ]
}
```

```js
// pages/some-page/+config.js

import HeadOpenGraph from './HeadOpenGraph'

// This page inherits HeadGlobal but not HeadDefaultOpenGraph

export default {
  Head: {
    value: HeadOpenGraph,
    group: 'open-graph',
    // Alternatively we can add `default: true` to the HeadDefaultOpenGraph value above
    clear: true
  }
}
```

**Comment by bighorik** [comment]
I would like to be able to flexibly configure this inheritance.
In my practice, a situation often happens when, for example, there are three pages:
applications/listing
applications/favorite
applications/create

For the first two, you need to use a common complex Layout, for the latter, a simpler layout. Accordingly, I would like to be able to set my own Layout for /create, overwriting the original one.

However, there are also reverse situations when you need to specify the Layout for 1 of the many pages, for example:
personalAccount/profile
personalAccount/settings
personalAccount/companies/portfolio
personalAccount/companies/stats
personalAccount/companies/about

In this case, I have added additional navigation for pages /companies - the ability to inherit previous Layouts is great

**Comment by brillout** [comment]
@bighorik Yeah, currently the only way is to do this:

```
applications/(complex-layout)/listing
applications/(complex-layout)/favorite
applications/(simple-layout)/create
```

Let me know whether it's a showstopper for you and I'll bump the priority of this ticket.

**Comment by AgentEnder** [comment]
I'd love a way to "escape" from a parent layout defined in the renderer. In my app, I have 

```
pages/{a bunch of pages}
pages/presentations/view
renderer/+Layout.jsx
```

`pages/presentations/view` needs to not have a surrounding layout, and currently I don't see a way to get it to not be present.

**Comment by brillout** [comment]
@AgentEnder Makes sense. I think this would be nice:

```js
// pages/some-page/+config.js

export default {
  // Remove all inherited layouts
  Layout: null
}
```

**Comment by AgentEnder** [comment]
That could work, currently I divide my PageShell component into MinimumPageShell and the base PageShell as saw above. Ideally I'd be able to set a Layout that is just the MinimumPageShell

**Comment by bighorik** [comment]
> @bighorik Да, в настоящее время единственный способ - сделать это:
> 
> ```
> applications/(complex-layout)/listing
> applications/(complex-layout)/favorite
> applications/(simple-layout)/create
> ```
> 
> Дайте мне знать, подходит ли вам это для показа, и я увеличу приоритет этого запроса.

Unfortunately, this is not suitable. I have a layout common to all pages, which at some levels of nesting of the file system must be completely overwritten. I'm using vike version 0.4.171 in production, and the lack of backward compatibility makes me cry :(

**Comment by brillout** [comment]
> Unfortunately, this is not suitable. I have a layout common to all pages, which at some levels of nesting of the file system must be completely overwritten.

Can you elaborate?

> I'm using vike version 0.4.171 in production, and the lack of backward compatibility makes me cry :(

The `0.4.x` release branch is backwards compatible. If it isn't then I consider it a bug; elaborate if that's the case.

**Comment by brillout** [comment]
Labeling this as `high-prio`. I understand it's annoying, but I'd like to gauge *how much* and *how often* it's annoying in order to prioritize accordingly. The more I understand how annoying it is the more I will prioritize this.

**Comment by chrisvander** [comment]
This is annoying for me too. I have a layout with a sidebar and a right-hand panel that *generally* is centered with a maximum width, but *occasionally* needs to be broken out to take up the full right hand width. Currently, I would need to redefine the typical case on every page, and only occasionally leave out the layout so the child can take up the full width.

I bet there are other cases like this, but I wonder how the default and overrides should really work. There's definitely parts of the layout I want on every page, parts that I want on most pages, and parts that I want on some pages. There's a hierarchy there and I'm not sure whether it's defaults or overrides or some other way to configure the layout.

There are also cases where I want to swap out portions of the sidebar depending on the context, and not have that be dynamic, but just a part of those page's layout.

**Comment by brillout** [comment]
Indeed, using [custom settings](https://vike.dev/meta) can be a workaround. Although note that all Layout code is loaded regardless of whether it's used then.

**Comment by brillout** [comment]
Bumping to `higest-prio` (although it's the last prio among the highest prio).

> I understand it's annoying, but I'd like to gauge _how much_ and _how often_ it's annoying in order to prioritize accordingly. The more I understand how annoying it is the more I will prioritize this.

Comment welcome. I could, maybe, implement this within a day.

**Comment by cr7pt0gr4ph7** [comment]
After digging through the source code of `vike` and giving it some time, I think there's a better & more generic solution that can be achieved by introducing an `inherit` config setting that controls which properties (accumuluated or not) are inherited from parent folders:

```ts
// src/pages/+config.ts
export default {
  Layout: DefaultLayout
}

// src/pages/login/+config.ts
export default {
  inherit: {
    Layout: false,
  },
  Layout: LoginLayout
}

// src/pages/article/seller-statistics/+config.ts
const AdminDefaults = {
  inherit: {
    Layout: false,
    title: false
  },
  Layout: [DefaultLayout, AdminLayout]
};

export default {
  extends: AdminDefaults
}


// src/pages/cart/checkout/+config.ts
import MinimalLayoutDefaults from 'pages/(minimal-layout)/defaults';

export default {
  extends: MinimalLayoutDefaults,
  Layout: CheckoutLayout
}

// src/pages/(minimal-layout)/+config.ts
export default {
  inherit: {
    Layout: false,
  },
  Layout: MinimalLayout
}

// src/pages/(minimal-layout)/defaults.ts
export default const MinimalLayoutDefaults = {
  inherit: {
    Layout: false
  },
  Layout: MinimalLayout
}
```

**Comment by drew-dulgar** [comment]
I am also interested in this inherited layout behavior. One thing that really caused some interesting behaviors is that my _errors/+Page.js file is also inheriting the default layout, which it shouldn't be. I had to basically move all of my base routes into a (app) folder to prevent errors from using the default layout. 

I get the entire Vike stack in a bit of a recursive trace because if something errors out in the default layout, it calls the error page, which calls the default layout with an error again. 

+1 for the inherits example above. I read over the initial suggestion and I'm not sure I quite understand the expected behavior with default/overrides option. It seems like it covers the needed scenarios, but it's not all that clear. I probably just need to mull it over.

**Comment by brillout** [comment]
(I'll have a closer look at your proposal once I'm done with what I'm currently working on. ETA next week.)

**Comment by k9982874** [comment]
I really need this option. In my app, some subpages won't inherit the layout from the parent.
I can't implement this gracefully at the moment.

**Comment by alexmnv** [comment]
@k9982874  For now, you can use a simple workaround: just define your own config setting for Layout rendering, which will not be cumulative.
Let's call it `PageLayout`:

```
meta: {
	PageLayout: {
		env: {server: true, client: true},
		cumulative: false,
	},
}
```

You can then render `PageLayout` inside the base layout. Here's an example using Vue:

```	
<script setup lang="ts">
import {usePageContext} from 'vike-vue/usePageContext';

const pageContext = usePageContext();
const PageLayout = pageContext.config.PageLayout;
</script>

<template>
	<PageLayout>
		<slot />
	</PageLayout>
</template>
```

This approach allows you to define `PageLayout` separately for each page (without cumulation)

**Comment by cr7pt0gr4ph7** [comment]
@brillout After thinking about it a bit more especially with respect to composability, and taking @alexmnv's suggestions as well as the applicability to other settings like `head` into account, I now see the value of your original proposal.

When I write `{ myConfigSetting: ... }`, my intent is to do one of four things:
- **Overwrite** _any inherited value_ completely
- **Append** to _array of inherited values_
- **Replace** a specific earlier element in _array of inherited values_
- **Merge** _new value_ with _inherited value_ in some custom fashion
  - This includes merging objects (think of `Object.assign`), because one somehow has to specify how sub-properties should be merged.

Values at higher hierarchy levels usually have a _default intent_, i.e. what should happen hen a child page specifies the same setting:

|               | Array-like<br>(e.g. `Head`, `Layout`)   | Array-like _default_<br>(e.g. `Layout`<br>sometimes?) | Object-like<br>(e.g. `meta`) | Primitive-like<br>(e.g. `title`) |
| ------------- | --------------------------------------- | ----------------------------------------------------- | ---------------------------- | -------------------------------- |
| **Overwrite** | :heavy_check_mark:                      | :heavy_check_mark:                                    | :heavy_check_mark:           | :white_check_mark:               |
| **Append**    | :white_check_mark:                      | :heavy_check_mark:                                    | :x: (makes no sense)         | :x: (makes no sense)             |
| **Replace**   | :heavy_check_mark:                      | :white_check_mark: (replace this,<br>keep inherited)  | :heavy_check_mark:           | (Same as _Overwrite_)[^2]        |
| **Merge**     | :heavy_check_mark: (but not useful)[^3] | :heavy_check_mark: (but not useful)[^3]               | :white_check_mark:[^4]       | :heavy_check_mark:[^5]           |

- :white_check_mark: Default intent
- :heavy_check_mark: Possible alternative intent

[^2]: _Replace_ is the same as _Overwrite_ for primitive values, because there is only ever one inherited value available.

[^3]: _Overwrite_, _Append_ and _Replace_ already cover all useful behaviors for _Array-like_ settings. Alternative stuff like also reordering inherited values is possible, but probably does not make sense from a readability/composability standpoint.

[^4]: Default merge behavior for properties of _Object-like_ settings has to be specified recursively.

[^5]: Although this can also be implemented as separate config settings plus a computed setting.

**Comment by brillout** [comment]
> (I'll have a closer look at your proposal once I'm done with what I'm currently working on. ETA next week.)

Tomorrow / beginning of next week. (Thanks for writing and thinking about all this! Looking forward to it.)

**Comment by chrisvander** [comment]
I'm curious how essential this is. I'm finding that using custom settings works great for the more complex use cases. What parts would be best to leave to the individual implementation?

**Comment by brillout** [comment]
Indeed, the proposed design can already be used today by reproducing it with a custom setting. It's just about improving the DX. It isn't about unblocking users. That said, it's a very a recurrent question so having it built-in would be nice.

For edge cases, it's probably better to teach users instead of having something built-in. (Actually, I'm realizing maybe we should teach how to implementing grouping instead of having it built-in. I guess we can skip grouping for now and making it built-in later if there is frequent demand for it.)

**Comment by chrisvander** [comment]
To me, it seems that the most pressing need may be addressed by just the inherit punch-out option. That would be a smaller feature that could be expanded on if there's a better DX. If someone wants to do a more sophisticated layout, they could compose by punching out and then importing the appropriate components.

**Comment by brillout** [comment]
Good point. Let's for now only implement `inherit: boolean` + teach about advanced use cases.

**Comment by brillout** [comment]
> Looking forward to settle on a design. It's indeed an essential feature that we should implement soon.

I guess the design is settled: I'll implement a new `inherit` option.

```js
// pages/admin-panel/+config.js

import { Layout } from './Layout'

export default {
  // Overrides all previously defined layouts (all layouts defined by an ancestor configuration)
  Layout: {
    value: Layout,
    inherit: false
  }
}
```

```js
// pages/+config.js

import { Layout } from './Layout'

export default {
  // Completely opt-out of Layout inheritance: a +Layout always overrides all previously defined +Layout
  Layout: {
    inherit: false
  }
}
```

Let me know if this doesn't work for your use case.

**Comment by brillout** [comment]
I'm considering implementing this, but I've one open question left. It would be nice to have `+Layout.inherit.js` similar to [`+data.{server,client,shared}.js`](https://vike.dev/data#environment) because, otherwise, it's a bit of a pain to have to create a `+config.js` file just to define `inherit: false`.

But we need the user to *opt-out* of inheritance, so I'm thinking one of the following options:
 - `+Layout.noInherit.js`
 - `+Layout.!inherit.js` <= would be my favorite but unfortunately doesn't seem to play nice with Linux
 - `+Layout.override.js` <= renaming `inherit: false` to `override: true`

**Comment by brillout** [comment]
Or maybe `+Layout.-inherit.js`. This should work on all filesystems. Con: looks a bit ugly, pro: it's clear (once you know what it means). Hm, not sure.

**Comment by cr7pt0gr4ph7** [comment]
When talking only about property names, I prefer `inherit: false` over `override: true`, as the former in my opinion makes it more clear _what_ is being overridden.

+1 for `+Layout.override.js`, as that names probably makes the most sense even one doesn't yet know about the inheritance mechanics. The naming of every other name involving some symbols like `~`, `^`, `!` is unclear anyway and requires looking up the documentation...

...so although having `.override.js` translate into `inherit: false` is not immediately obvious, at least the two variants by themselves are self-teaching.

**Comment by brillout** [comment]
@cr7pt0gr4ph7 How about `+Layout.noInherit.js`?

So far I'm hesitating between `+Layout.noInherit.js` and `+Layout.-inherit.js`. If I imagine myself using `+Layout.-inherit.js` regularly then I think I would get used to it and it would fit nicely, but it does look a bit cryptic 😅

**Comment by cr7pt0gr4ph7** [comment]
`+Layout.noInherit.js` (suffix matching should probably be case-insensitive and also support `+Layout.noinherit.js`) would be my second choice after `+Layout.override.js`, and I'd also consider that an acceptable tradeoff, though I personally prefer `+Layout.override.js`.

`+Layout.-inherit.js` looks cryptic to me, and IMO it's too visually similar to `+Layout.inherit.js`, which makes accidental typos harder to spot.

**Comment by brillout** [comment]
FYI the reason I prefer `no`/`-` over `override` is because it can be re-used for other configs.

**Comment by AgentEnder** [comment]
I don't like the `-inherit` syntax, but `noInherit` or `no-inherit` would both be fine imho

**Comment by enri90** [comment]
Has a decision been made? I don't understand if it is already possible to do so

**Comment by brillout** [comment]
Yes, unless someones argues against I'll go for `+Layout.noInherit.js` (or setting `Layout.inherit: false` inside `+config.js`).

In principle it should be easy to implement (settling on the design was the difficult part).

@enri90 Is it a blocker for you? If it is then I'll have a look at implementing it in the coming days.

**Comment by shahyar** [comment]
Would love to get this implemented soon!

**Comment by AgentEnder** [comment]
Potentially separate, it'd be useful to be able to make settings that aren't normally inherited, inherited. E.g. nested pages may all need the sane data

**Comment by brillout** [comment]
@AgentEnder Yes that's very much the plan. E.g. you'll be able to make `+data` cumulative (which seems to be what you want).

**Comment by brillout** [comment]
With Dani we're discussing the implementation. I think we found an even better name instead of `.noInherit.js`:

```yaml
pages/admin-panel/+Layout.clear.js   # Clear all inherited values
pages/marketing/+Layout.default.js   # Any child +Layout.js will override it
````

**Comment by brillout** [comment]
Implemented by @nitedani and pre-released as `0.4.237-commit-92dc549`.

**Comment by AgentEnder** [comment]
Oh, awesome! I just hit this again and saw the release. Interestingly when playing around with it though, I don't think the typings in config have been updated as it doesn't appear to support the object syntax from the issue description from a typescript perspective but it also doesn't appear to work in the config file. The `+Layout.clear.ts` worked fine though!

@brillout / @nitedani  is this expected?

**Comment by brillout** [comment]
Only the filename suffices are supported (see updated docs). Feel free to let us know if you need the meta object syntax.
